### DISCUSSION TITLE
[Feature Request] Unification of Email processing

### DISCUSSION THREAD

**Discussion by madduck** [paperless-ngx/paperless-ngx#6929] [Feature Requests]
### Description

At the moment, Paperless supports slurping emails via IMAP, as well as the import of `.eml` files. But the two seem to happen quite independently and differently. While the first generally also adds all attachments of a message to the document set (see https://github.com/paperless-ngx/paperless-ngx/discussions/6834), the import of `.eml` files ignores all attachments and only produces a PDF view of the message itself.

Those two should be unified, and email processing should be the same no matter how a message enters the system.

Do you agree?

**Comment by madduck** [comment]
I am willing to work on this, but I don't want to invest time that ends up being wasted, i.e. when a PR eventually receives as much engagement as this simple question?

**Reply by rqi14** [reply]
> I am willing to work on this, but I don't want to invest time that ends up being wasted, i.e. when a PR eventually receives as much engagement as this simple question?

I'm using paperless to automatically collect CVs for recruitment. Often people add short introduction of themselves in the body of the email with their CVs attached. It will be very helpful if one can see both the body and the attachment of emails. It does make a lot more sense than saving separate files.

**Comment by SmilingGit** [comment]
Great idea, I fully agree! @shamoon (thank you!) noticed that I posted a similar request to yours here:

https://github.com/paperless-ngx/paperless-ngx/discussions/10548

I could live with something simpler than a unification, but indeed I agree this would be the best solution.

**Comment by harald-l** [comment]
To give another example, where unification would be useful: Encrypted .eml files.

If your Paperless-ngx instance is not connected to the internet and/or [setup with gpg agent](https://docs.paperless-ngx.com/advanced_usage/#gpg-decryptor) not feasible, best solution is to just manually import .eml files. This also has the advantage, that only relevant mails can be selectively imported, no need for IMAP account access to all mails.

Rough Workflow:
1. Check mails with client of choice - e.g. Thunderbird, which also supports GPG
2. Selectively decrypt all relevant/important mails - Thunderbird can easily do this via "Decrypt into folder"
3. Import decrypted `.eml` files into Paperless-ngx (either manually via upload or import folder), incl. optional attachments.

---

Thanks @madduck for your effort so far!
> I am willing to work on this, but I don't want to invest time that ends up being wasted

Are you still waiting for feedback by maintainers?

**Comment by madduck** [comment]
@harald-l i started on this, but it does not help my motivation that I have to do this isolatedly...

**Reply by SmilingGit** [reply]
Unfortunately, I'm not experienced enough to provide help with coding, but I'm happy to help with testing and debugging!

**Reply by harald-l** [reply]
>  started on this, but it does not help my motivation that I have to do this isolatedly...

Ah OK, just being curious. Hopefully you did not perceive above comment in a sense, that this feature would be your own burden or responsibility. I'd like to help as well - unfortunately I am not familiar with Paperless-ngx codebase (yet). But If you got any pull request draft, I certainly could have a look at it. Cheers

**Reply by SmilingGit** [reply]
@madduck — just curious to know if you’ve had the chance to get any further. I’m still happy to help with debugging and testing, or anything else. All the best!

**Comment by madduck** [comment]
I am not making progress. I need some guidance, but the devs have so far simply ignored all requests. Plus, given how most issues and discussions just auto-lock/-close without any engagement, I am weary to invest time into this project for fear that it'll just get ignored. I am not interested in maintaining a fork.

**Reply by shamoon** [reply]
Such attitude.

Paperless-ngx receives thousands of issues and discussions. Not every request is going to get engagement, and the level of response a post receives is usually a reflection of community interest, not somehow of negligence by us.

Comments suggesting that the project “ignores requests” are inaccurate and unhelpful. This is an open-source project, and we expect contributors to approach it constructively and respectfully.

Given the tone of your comments and past interactions, if you feel your time is being wasted here, then I agree it may be best for you not to contribute further.

**Comment by stratus-ss** [comment]
@shamoon 

I have some decent background in python. I may be able to help out with this. However, unifying code bases can come down to a matter of style and opinion (similar to general refactoring).

Do you have any opinions on how in general I might approach this? Any guidance to start me in the right direction might be useful.

Thanks

**Reply by shamoon** [reply]
The paperless codebase is complex so I would start by looking to understand what is described and come up with a plan before executing it. In some ways I don’t think it would be so hard. I will take a look myself today.

Only other thing to be aware of is that we are planning some larger changes in our next major release (spearheaded by our main backend dev stumpylog) so if the changes here turn out to be very significant it might not make sense to implement before that.

**Reply by madduck** [reply]
Thank you for your response! Do you know why there are two code path in the first place?

**Reply by stratus-ss** [reply]
if we have a large code base shift, it might be ill-advised to start an undertaking like this right now depending on how far out (ish) the new code base is.

Is there a rough eta @shamoon

**Reply by shamoon** [reply]
No ETA. We'll be releasing 2.19 soon-ish, then likely some bug fix releases and then working towards v3. Given v3 will likely include larger changes, my guess is it will "be a while". A change to this logic wouldn't necessarily need to wait until v3 is "ready", just until development has begun in earnest to avoid wasted effort. But up to you of course.

**Reply by stratus-ss** [reply]
thanks for the reply. For me, I was going to help this along for a friend. I am not using this feature myself

**Reply by madduck** [reply]
Is there a roadmap? Maybe this code is best done as part of major changes anyway?

**Comment by samoht62** [comment]
I am supporting the orginal request. To connect paperless to my email account makes no sense for me as app. 0.1% of the incoming emails are worth to be stored. So it would ne really helpful to save .eml with attachments to the consume directory and get email and attachment processed.
Thanks in advance realising this feature.

**Comment by GeorgeCastanza** [comment]
Same.  I support this too.

**Comment by madduck** [comment]
@samoht62 @GeorgeCastanza please make sure to upvote the discussion itself!

**Comment by sebastian-de-chiara** [comment]
This would be an incredible addition to paperless. Having both email text and attachments in one single file would be extremely useful. Pretty please consider it!

**Comment by mikebakke** [comment]
100% this is a good request. I'm saving tutanota emails and they do not allow external connection to their servers so drag and drop is currently the only way. 

I guess the positive is that the attachment does seem to be ingested as part of the .eml and is restored when the document is downloaded.  It just seems we're *so* close to having a seamless experience. I very much hope this can be addressed in V3

**Comment by randolph005** [comment]
Do I perceive correctly, there is no offline way for paperless-ngx to receive and fully process E-Mails? I.e. You need  an online connection to the E-Mail account currently?
 
Our paperless-ngx instance has no internet access. Is there a workaround to process `.eml` incl. attachments by just sending the E-Mail message as plain `.eml` to server?

**Comment by SafetyIng** [comment]
Hey @madduck , has there been any progress? Is there a way to support this?

Would you consider taking into account the behavior I described in #13030 as well?

**Comment by madduck** [comment]
@SafetyIng I have not made progress on this, largely because I am awaiting v3 to stabilise.
